Userland/Libraries/LibJS/SourceCode.cpp

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2022, Andreas Kling <kling@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Utf8View.h>
+#include <LibJS/SourceCode.h>
+#include <LibJS/SourceRange.h>
+
+namespace JS {
+
+NonnullRefPtr<SourceCode> SourceCode::create(String filename, String code)
+{
+    return adopt_ref(*new SourceCode(move(filename), move(code)));
+}
+
+SourceCode::SourceCode(String filename, String code)
+    : m_filename(move(filename))
+    , m_code(move(code))
+{
+}
+
+String const& SourceCode::filename() const
+{
+    return m_filename;
+}
+
+String const& SourceCode::code() const
+{
+    return m_code;
+}
+
+SourceRange SourceCode::range_from_offsets(u32 start_offset, u32 end_offset) const
+{
+    Position start;
+    Position end;
+
+    size_t line = 1;
+    size_t column = 1;
+
+    Utf8View view(m_code);
+    for (auto it = view.begin(); it != view.end(); ++it) {
+
+        if (start_offset == view.byte_offset_of(it)) {
+            start = Position {
+                .line = line,
+                .column = column,
+                .offset = start_offset,
+            };
+        }
+
+        if (end_offset == view.byte_offset_of(it)) {
+            end = Position {
+                .line = line,
+                .column = column,
+                .offset = end_offset,
+            };
+            break;
+        }
+
+        if (*it == '\n') {
+            ++line;
+            column = 1;
+            continue;
+        }
```

Comment:
This will not quite give the same results as currently as the are some valid unicode code points we also need to consider as line endings.
See:
https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibJS/Lexer.cpp#L345-L354

You'll also need to be careful with `'\r'` as a `'\r\n'` sequence should count as just one.

Reply:
Ooh, great catch! Will fix.